src/nf/nf_optimizers.f90

```diff
@@ -123,4 +135,47 @@ pure subroutine minimize_rmsprop(self, param, gradient)
 
   end subroutine minimize_rmsprop
 
+
+  impure elemental subroutine init_adam(self, num_params)
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
+      allocate(self % m(num_params))
+      self % m = 0
+    end if
+    if (.not. allocated(self % v)) then
+      allocate(self % v(num_params))
+      self % v = 0
+    end if
```

Comment:
`m` and `v` will always be either both allocated or deallocated so we can fold the second `if` block into the first one.

```suggestion
    if (.not. allocated(self % m)) then
      allocate(self % m(num_params), self % v(num_params))
      self % m = 0
      self % v = 0
    end if
```